posts/2021-04-21-What I wish somebody told me when I was learning Haskell.md

```diff
@@ -0,0 +1,337 @@
+
+I hope that below ideas could help reducing this investment - some of them helped me, and some I had to figure out in a hard way.
+
+## How to read and think about code
+
+### Read types first, not code
+
+We take it for granted that to understand "what the program does" you have to "read its code". Unfortunately, coming from imperative languages, we do not see types as part of the code - we see them as annotations, something secondary, being there to appease compiler.
+
+This mindset does not work in Haskell. Trying to understand what the function does purely from its code is at least ineffective, and in many cases it is not possible.
+
+To get fluent and productive with Haskell you should not try to understand what a function "does", you should aim to understand what it "is" by looking at its type first, and to start seeing the types as the primary part of your code that drives, rather than restricts, your implementation.
+
+Once you get into the habit of reading the types first, you are likely to discover that most libraries have more than enough documentation to use them effectively, contrary to a popular belief.
+
+### Read it slowly
+
+To become effective at reading Haskell you have to unlearn how you read imperative languages.
+
+The problem for the beginners is that we are used to having a lot of code that has to be read as prose. Haskell code is can be very terse, and it has to be read slower, as if it were a formula in some cases. When you get used to it, you can write and read the same logic much faster - and it also makes it easier to understand it later.
+
+I had a particular example when 70 lines of TypeScript code took 8 one-line functions in Haskell (and all lines were normal length and quite readable), so 16 lines together with type signatures [^dict].
+
+Reading each Haskell line can take twice longer, but because of much smaller overall size it is at least twice faster to write and to read than the equivalent logic in TypeScript.
+
+### Avoid terms used in imperative languages
+
+It may seem insignificant, but using the correct terminology would help switching to the functional mindset:
+
+- functions are not "called" and "executed", they are "used" and "evaluated".
+- functions do not "return" values, even when they use `return` function, they are a binding of function code to the function name.
+- preferring `pure` over `return` [^pure] would both save you typing and help switching mindset.
+- "names" are not "variables", they are not "assigned", they are "bound" to values.
+- `do` syntax does not define execution steps, it defines the sequence of actions bound via their Monad type class interface.
+
+## Model your program with types
+
+### Types are are the model of your program
+
+Types in Haskell are not just annotations - they are the core of the language used to model the program and to define the shape of and constraints on the actions that the program will perform.
```

Comment:
This sentence is somewhat cumbersome to read, particularly due to "shape of and constraints on" and it being way far in sentence structure.. Maybe just the following?

> ... used to model the program and to define its shape and constraints

Definitely less eloquently, but loosely speaking the program kind of is the actions being performed, so not much meaning is lost

Reply:
or
> ... and to define shape and constraints for actions ... / for its actions